[WIKI] Install - Rewrite "Installing Loris in Depth" for new Wiki (#5253)
aces/Loris: docs/wiki/00 - SERVER INSTALL AND CONFIGURATION/01 - LORIS Install/Ubuntu/README.md

# Installing LORIS In-Depth

This guide assumes an installation on an Ubuntu server. 

If you are using CentOS, please visit the CentOS installation guide for 
instructions on installation.

We also have instructions for how to set up LORIS for macOS. However, only
 Ubuntu and CentOS are officially supported by the team so we may be unable to
 help you with issues arising on other systems.

> 

Reviewer: ```suggestion
```

We shouldn't be pointing people to things we're explicitly not supporting.

When you've completed this guide, you should be able to load and log into LORIS in your browser. Further setup to customize LORIS for your project will be required after that. Please visit
[the Setup page](https://github.com/aces/Loris/wiki/Setup) in order to complete the setup for LORIS.

## Prequisities

### Server Stack

LORIS requires a LAMP stack in order to run, specifically:
* Apache 2.4
* MySQL 5.7 (or MariaDB 10.3) (or higher)
* PHP 7.3 (or higher)

> Additionally, LORIS requires the following package managers:

Reviewer: LORIS doesn't require these, building LORIS does.

> * npm 8.0 (or higher)

Reviewer 2: @johnsaigle I think this should say Node 8.0
It looks like npm 's release numbers are not that high. 
(reported via Blake Fitch in Tuebingen)

* composer

This guide does not cover installation of these requirements.

### Apt Packages
The following Ubuntu packages are required and should be installed using 
`sudo apt install ...`.

> * git

Reviewer: The description below is using a tarball, there should be nowhere that git is required.

* curl

> * wget

Reviewer: The only reason `curl` and `wget` are required is because you're using them in the command below. Why are you using both? Choose one (probably curl) and stick with it..

* php7.3-xml
* php7.3-mbstring
* php7.3-gd
* php7.3-zip
* libapache2-mod-php7.3

## Getting the source code

Visit the [releases](github.com/aces/loris/releases) page and download the zipped file of the latest LORIS
release. The below command will download the latest LORIS release and save 
it to a file named `loris-src.tar.gz`.

```bash
curl -s https://api.github.com/repos/aces/loris/releases/latest \
| grep 'tarball' \
| cut -d : -f2,3 \
| tr -d \" \
| tr -d , \
| wget -i - -O loris-src.tar.gz
```

When this is complete, expand the compressed file and move its contents to the
directory from which LORIS will be served. 

```bash
tar -zxf loris-src.tar.gz
```

This will create a folder called something like `aces-Loris-9e30cf0`. (The
part after `aces-Loris` is not important).

Rename this file to `loris`, e.g.

```bash
mv aces-Loris-9e30cf0 loris
```

Then move this folder to the web root (e.g. `/var/www/`) and go to this 
directory.

```bash
mv ./loris /var/www/; cd /var/www/loris

### If installing LORIS for production

```bash
cd `/var/www/loris/`; make
```

## Running the install script

The next step in setting up LORIS is running the script `install.sh` in the 

> `tools/` directory. The script must be run from this directory.

Reviewer 3: ```suggestion
`tools/` directory. The script must be run from that directory.
```

This will begin an interactive setup process that will configure files and
permissions required to get LORIS up and running.

The install script will prompt for the user password in order to use `sudo` permission for configuring
certain file permissions as well as Apache. However, the script should not
be run as `root` and will exit automatically if it detects that it has been
executed with superuser permissions.

```bash
cd tools/
./install.sh
```

## Configuring the database

The install script will tell you to navigate to <loris-url>/installdb.php.

MySQL (or MariaDB) must be installed and a root or admin-level MySQL user must
be created before continuing. (This is not the same as a unix root credential.)

This web page will prompt you for the following information:

>  * `Server Hostname`. Use `localhost` if your database is hosted on your locally. Use the IP address of your database server otherwise.

Reviewer 3: ```suggestion
 * `Server Hostname`. Use `localhost` if your database is hosted on the same machine as your web server. Use the IP address of your database server otherwise.
```
or something else that make more sense then current

# Installing LORIS In-Depth

This guide assumes an installation on an Ubuntu server. 

If you are using CentOS, please visit the CentOS installation guide for 
instructions on installation.

We also have instructions for how to set up LORIS for macOS. However, only
 Ubuntu and CentOS are officially supported by the team so we may be unable to
 help you with issues arising on other systems.

> When you've completed this guide, you should be able to load and log into LORIS in your browser.  Further setup to customize LORIS for your project will be required after tha. Please visit

Reviewer 3: ```suggestion
When you've completed this guide, you should be able to load and log into LORIS in your browser. Further setup to customize LORIS for your project will be required after tha. Please visit
```

* php7.3-xml
* php7.3-mbstring
* php7.3-gd
* libapache2-mod-php7.3

## Getting the source code

### LORIS for Development
If installing LORIS for development purposes, clone the source code from our
GitHub repository to the web root. Then change to this directory.

Example command assuming LORIS will be served from `/var/www/`:

```bash
git clone git@github.com:aces/Loris.git /var/www/loris
cd /var/www/loris/
```

### LORIS for Production

Visit the [releases](github.com/aces/loris/releases) page and download the zipped file of the latest LORIS
release. The below command will download the latest LORIS release and save 
it to a file named `loris-src.tar.gz`.

```bash
curl -s https://api.github.com/repos/aces/loris/releases/latest \
| grep 'tarball' \
| cut -d : -f2,3 \
| tr -d \" \
| tr -d , \
| wget -i - -O loris-src.tar.gz
```

When this is complete, expand the compressed file and move its contents to the
directory from which LORIS will be served. 

```bash
tar -zxf loris-src.tar.gz
```

This will create a folder called something like `aces-Loris-9e30cf0`. (The
part after `aces-Loris` is not important).

Rename this file to `loris`, e.g.

```bash
mv aces-Loris-9e30cf0 loris
```

> Then move this folder to the web root (e.g. `/var/www/`) and go to this 

Author: I've added the command below.

> directory.

Author: ```suggestion
directory.

```bash
mv ./loris /var/www/; cd /var/www/loris
```

Example command assuming LORIS will be served from `/var/www/`:

```bash
git clone git@github.com:aces/Loris.git /var/www/loris
cd /var/www/loris/
```

### LORIS for Production

Visit the [releases](github.com/aces/loris/releases) page and download the zipped file of the latest LORIS
release. The below command will download the latest LORIS release and save 
it to a file named `loris-src.tar.gz`.

```bash
curl -s https://api.github.com/repos/aces/loris/releases/latest \
| grep 'tarball' \
| cut -d : -f2,3 \
| tr -d \" \
| tr -d , \
| wget -i - -O loris-src.tar.gz
```

When this is complete, expand the compressed file and move its contents to the
directory from which LORIS will be served. 

```bash
tar -zxf loris-src.tar.gz
```

This will create a folder called something like `aces-Loris-9e30cf0`. (The
part after `aces-Loris` is not important).

Rename this file to `loris`, e.g.

```bash
mv aces-Loris-9e30cf0 loris
```

Then move this folder to the web root (e.g. `/var/www/`) and go to this 
directory.

```bash
mv ./loris /var/www/; cd /var/www/loris

### If installing LORIS for production

```bash
cd `/var/www/loris/`; make
```

> ### If installing LORIS for development

Reviewer: I think it's confusing that this section is here and not in the "Development" section below.. following these instructions will also result in a "Development" environment that was installed from a tarball and isn't tracked by git. I think this document should simply have an audience and stick to it rather than trying to provide different details to different audiences.

Author: Agreed I'll refactor it

MySQL (or MariaDB) must be installed and a root or admin-level MySQL user must
be created before continuing. (This is not the same as a unix root credential.)

This web page will prompt you for the following information:

 * `Server Hostname`. Use `localhost` if your database is hosted on your locally. Use the IP address of your database server otherwise.

 * `Admin Username` A database user with permission to create databases and tables.

 * `Admin Password` The password for the above database user.

 * `Database Name` Defaults to "LORIS".

Click submit, and on the next screen that is presented, follow instructions to enter the username and password of your LORIS database user and front-end `admin` user.

If you encounter issues creating/generating your config file, you may have to manually paste the xml output that appears on the screen into the file `/var/www/loris/project/config.xml`

Your LORIS instance should now be accessible by pointing your browser URL to `http://%IPADDRESS%`.

Now that the installation is complete, follow the [Setup process](https://github.com/aces/Loris/wiki/Setup) to customize your project.

### Development

> 

Reviewer 2: Suggestion - Given there's a Development section here, we could/should mention setting the config.xml parameter here (and if not here, in the following Setup step), and anything you might do differently in dev/production.

Author: Since that information doesn't depend on the Operating System, I think it should be done in another document.

# Installing LORIS In-Depth

This guide assumes an installation on an Ubuntu server. 

If you are using CentOS, please visit the CentOS installation guide for 
instructions on installation.

> We also have instructions for how to set up LORIS for macOS. However, only

Reviewer 2: add macos link

Author: https://github.com/aces/Loris/pull/5645 is changing the name and location of this file. We can amend this document to link to its new location later on.

 Ubuntu and CentOS are officially supported by the team so we may be unable to
 help you with issues arising on other systems.

> 

Reviewer 2: remove extra blank line

> When you've completed this guide, further set up is required. Please visit

Reviewer 2: Let's spell out the scope of this page a little bit more:
```suggestion
When you've completed this guide, you should be able to load and log into LORIS in your browser.  Further setup to customize LORIS for your project will be required after that -Please visit
```

[the Setup page](https://github.com/aces/Loris/wiki/Setup) in order to be
able to use LORIS properly.

## Prequisities

### Server Stack

LORIS requires a LAMP stack in order to run, specifically:
* Apache 2.4

> * MySQL 5.7 (or MariaDB 10.1) (or higher)

Reviewer 3: ```suggestion
* MySQL 5.7 (or MariaDB 10.3) (or higher)
```

* PHP 7.3 (or higher)

Additionally, LORIS requires the following package managers:
* npm 8.0 (or higher)
* composer

This guide does not cover installation of these requirements.

### Apt Packages
The following Ubuntu packages are required and should be installed using 
`sudo apt install ...`.

* git
* curl
* wget
* zip
* unzip

> * php-json

Reviewer 3: php-json and php7.3-json?

Author: Thanks I deleted `php73-json`.

Reviewer 3: shouldn't it be the opposite? keep php73-json.

Author: I tried running `php-json` on my VM and it started downloading php74-json on its own. i think it's a smart(er) parent package that takes care of the rest.

* make
* software-properties-common
* php7.3-mysql
* php7.3-xml

> * php7.3-json

Author: ```suggestion
```

`php-json` is correct

* php7.3-mbstring
* php7.3-gd
* libapache2-mod-php7.3

### Database set up

> MySQL (or MariaDB) must be installed and a root or admin-level MySQL user must

Reviewer 2: Because it's tremendously useful for support and troubleshooting, I think we should keep using the placeholder names for these accounts/roles: `lorisadmin` and `lorisdbadmin`.  
Is there a reason they're not here? 

By all means let's improve upon them if we can (e.g. maybe `dbadmin` because `lorisdbadmin` is too close to the unix user `lorisadmin`?)

Reviewer 2: Not really comfortable resolving this conversation, since i'd like these comments to be retained on the PR...

be created before continuing. (This is not the same as a unix root credential.)

## Getting the source code

### LORIS for Development
If installing LORIS for development purposes, clone the source code from our
GitHub repository to the web root. Then change to this directory.

Example command assuming LORIS will be served from `/var/www/`:

```bash
git clone git@github.com:aces/Loris.git /var/www/loris
cd /var/www/loris/
```

> 

Reviewer 2: add a note here something like: when documentation refers to the Loris root directory, it means `/var/www/loris`/

Author: I don't use that term in that way in this document and it's also not quite right. The LORIS root directory is the folder where all the LORIS files are contained. `/var/www/loris/` is one example name and path to a LORIS root directory. If someone installed LORIS at `/home/john/my_coding_projects/TestLoris` then 'TestLoris' is the LORIS root directory.

I also don't want this document to function as a glossary for example values for configuration settings. If that's useful we can create a separate document for that which other docs can link to. Otherwise we would have to duplicate this information across CentOS and macOS install guides, at least.

### LORIS for Production

> Visit the releases page and download the zipped file of the latest LORIS

Reviewer 2: ```suggestion
Visit the [releases](github.com/aces/loris/releases) page and download the zipped file of the latest LORIS
```

| tr -d \" \
| tr -d , \
| wget -i - -O loris-src.tar.gz
```

When this is complete, expand the compressed file and move its contents to the
directory from which LORIS will be served. 

```bash
tar -zxf loris-src.tar.gz
```

This will create a folder called something like `aces-Loris-9e30cf0`. (The
part after `aces-Loris` is not important).

Rename this file to `loris`, e.g.

```bash
mv aces-Loris-9e30cf0 loris
```

Then move this folder to the web root (e.g. `/var/www/`) and go to this 
directory.

### If installing LORIS for production

```bash
cd `/var/www/loris/`; make
```

### If installing LORIS for development

```bash
# This is necessary for 'phan', a static analysis tool
sudo pecl install ast-1.0.3
cd `/var/www/loris/`; make dev
```

The `dev` target for `make` will install additional libraries that are needed 
for development process but not for production installs.

## Running the install script

The next step in setting up LORIS is running the script `install.sh` in the 
`tools/` directory. The script must be run from this directory.

This will begin an interactive setup process that will configure files and
permissions required to get LORIS up and running.

> The install script will prompt for a `root` password in order to configure

Reviewer 3: ```suggestion
The install script will prompt for a the user password in order to use `sudo` permission for configuring
```
Look like in your rebase, you went a little to far back.

Reviewer 3: ```suggestion
The install script will prompt for the user password in order to use `sudo` permission for configuring
```
I hadn't cut at the right place.

 * `Admin Password` The password for the above database user.

 * `Database Name` set to "LORIS" by default, can be customised

Click submit, and on the next screen that is presented, follow instructions to enter the username and password of your LORIS database user and front-end `admin` user.

If you encounter issues creating/generating your config file, you may have to manually paste the xml output that appears on the screen into the file `/var/www/loris/project/config.xml`

Your LORIS instance should now be accessible by pointing your browser URL to `http://%IPADDRESS%`

If there are any errors or you get a blank page, troubleshoot the errors in your apache error log - by default
 `/var/log/apache2/loris-error.log`

## Configuring MySQL

Please [go here](https://github.com/aces/Loris/wiki/Hosting-the-Database-Myself)
for information on configuring MySQL.

### Development

The name of our database filled with dummy test data is "Raisinbread". This 
can be used by developers to populate LORIS with useable data for testing
changes to the codebase.

Instructions on installing Raisinbread can be found in the LORIS root folder
at `raisinbread/README.md`

## Troubleshooting

> If errors are encountered during the section "Running the install script",

Reviewer 2: I'd like to see here quick links to top 3 troubleshooting questions we get right after Install, which aren't script errors. i.e. What to do if : 
* front end doesn't load -> check/fix config paths via the mysql back-end
* css looks busted (if that is still happening) -> check/fix config paths via the mysql back-end
* you need to reset your admin password -> run the tools script

Author: I think we should abstract the troubleshooting document into its own file. These errors aren't specific to Ubuntu so they shouldn't be in the Ubuntu instructions.

Reviewer 2: looks like this is #5920 - new issue.

ensure that all the prequisites for LORIS are installed and meet the minimum
required versions.

> Ensure you've followed the [Setup process](https://github.com/aces/Loris/wiki/Setup)

Reviewer 2: Easier and clearer to read --
```suggestion
Next: follow the [Setup process](https://github.com/aces/Loris/wiki/Setup) to customize your LORIS.
```

# Installing LORIS In-Depth

This guide assumes an installation on an Ubuntu server. 

If you are using CentOS, please visit the CentOS installation guide for 
instructions on installation.

We also have instructions for how to set up LORIS for macOS. However, only
 Ubuntu and CentOS are officially supported by the team so we may be unable to
 help you with issues arising on other systems.

When you've completed this guide, further set up is required. Please visit
[the Setup page](https://github.com/aces/Loris/wiki/Setup) in order to be
able to use LORIS properly.

## Prequisities

### Server Stack

LORIS requires a LAMP stack in order to run, specifically:
* Apache 2.4
* MySQL 5.7 (or MariaDB 10.1) (or higher)
* PHP 7.2 (or higher)

Additionally, LORIS requires the following package managers:
* npm 8.0 (or higher)
* composer

This guide does not cover installation of these requirements.

### Apt Packages
The following Ubuntu packages are required and should be installed using 
`sudo apt install ...`.

* git
* curl
* wget
* zip
* unzip
* php-json
* make
* software-properties-common
* php7.2-mysql
* php7.2-xml
* php7.2-json
* php7.2-mbstring

> * php7.2-gd

Reviewer 4: Is php-curl not a dependency ?

Author: I'm not sure if it is... I think we're using Guzzle for requests to other domains instead of raw curl (https://github.com/aces/Loris/blob/master/src/Http/Client.php)

Reviewer 4: oh okai, nvm then

`tools/` directory. The script must be run from this directory.

This will begin an interactive setup process that will configure files and
permissions required to get LORIS up and running.

The install script will prompt for a `root` password in order to configure
certain file permissions as well as Apache. However, the script should not
be run as `root` and will exit automatically if it detects that it has been
executed with superuser permissions.

```bash
cd tools/
./install.sh
```

## Compiling source code and installing dependencies
Once this step is complete, run the `make` command to install additional
dependencies and compile JavaScript needed to properly render the LORIS
front-end.

`make` must be run from the LORIS root, not `tools/`

### Installing LORIS for production

```bash
cd `/var/www/loris/`; make
```

### Installing LORIS for development

```bash
# This is necessary for 'phan', a static analysis tool
sudo pecl install ast-1.0.3
cd `/var/www/loris/`; make dev
```

The `dev` target for `make` will install additional libraries that are needed 
for development process but not for production installs.

## Configuring the database

> Open your browser and go to: `http://%IPADDRESS%/installdb.php` where 

Reviewer 4: Why IP address ? It could be domain name

Author: True, I'll change this
